### DISCUSSION TITLE
Storybook: Export Library with PrimeVue with Aura theme customized

### DISCUSSION THREAD

**Discussion by cristinafernandez1234** [primefaces/community#2436] [PrimeVue]
We have a component library that uses PrimeVue components and custom components from our own, everything with our own design system.

Until PrimeVue 3 we were using a theme provided by PrimeVue that we customized: https://github.com/primefaces/primevue/blob/master/public/themes/saga-blue/theme.css.

Now, with PrimeVue 4, we are using the Aura theme and customize it to adapt to our design. Here a mini example:

```
import { definePreset } from "@primevue/themes";
import Aura from "@primevue/themes/aura";

// Aura preset (https://primevue.org/theming/styled/) adapted to project
const CustomPreset= definePreset(Aura, {
primitive: {
    teal: {
      100: "#edf7f7",
      500: "#008580",
      800: "#0c4543"
    }
  },
  semantic: {
    primary: {
      100: "{teal.100}",
      500: "{teal.500}",
      800: "{teal.800}"
    }
  },
  components: {
    button: {
      padding: {
        x: "16px"
      },
      label: {
        fontWeight: "700"
      },
      colorScheme: {
        light: {
          outlined: {
            primary: {
              hoverBackground: "{primary.100}",
              borderColor: "{primary.500}",
              color: "{primary.500}"
            }
          },
          text: {
            primary: {
              hoverBackground: "{primary.100}",
              activeBackground: "{primary.100}"
            }
          }
        }
      }
    }
});

export default CustomPreset;
```

We use it without problems in the preview.js of our storybook:

```
import CustomPreset from "../src/CustomPreset.js";

app.use(PrimeVue, {
    theme: {
      preset: CustomPreset
    }
  });
```

If we check in the console we see the CSS variables from Aura theme generated (example `--p-button-primary-background`): 

<img width="735" alt="storybook" src="https://github.com/user-attachments/assets/3f6db210-1097-4715-92f3-a7d0199bf4d4">

But the problem we have is that we have to export the library with the Aura style customized and use it in other projects. Before we exported the CSS customized, how can we achieve the same from the Aura theme? We don't see any documentation about it. 

We tried to export the CustomPreset file and uses in the Vue 3 project in main.js:

```
app.use(PrimeVue, {
    preset: CustomPreset
  });
```

But the `--p-button-primary-background` and the other vars are not defined there:

<img width="218" alt="styles" src="https://github.com/user-attachments/assets/79c895f1-d506-4673-bcf3-2d5a6de5046f">

Thanks in advance.

**Comment by j0rgedev** [comment]
Hello @cristinafernandez1234. I don't get to understand your use case, why do you want to use your CustomPreset with this code?
```javascript
app.use(PrimeVue, {
    preset: CustomPreset
  });
```

**Reply by cristinafernandez1234** [reply]
Hello @j0rgedev , sorry it would be with same structure as we used in the Storybook project:

```
app.use(PrimeVue, {
    theme: {
      preset: CustomPreset
    }
  });
```

To clarify:

We export the following index.js with a rollup.config.js:

```
import "./assets/style/global.css";

import PrimeVue from "primevue/config";
import ConfirmationService from "primevue/confirmationservice";
import ToastService from "primevue/toastservice";
import Tooltip from "primevue/tooltip";

import { FilterMatchMode, FilterService, FilterOperator } from "@primevue/core/api";
import { UniqueComponentId } from "@primevue/core/utils";

import { useToast } from "primevue/usetoast";

import CustomPresetfrom "./CustomPreset.js";

export {
  PrimeVue as DesignSystemConfig,
  ConfirmationService,
  ToastService,
  Tooltip,
  FilterMatchMode,
  FilterService,
  FilterOperator,
  UniqueComponentId,
  useToast,
  CustomPreset
};

```

Then we import it in a Vue 3 project in main.js:

```
import { createApp } from "vue";
import { createPinia } from "pinia";
import App from "./App.vue";
import router from "./router";
import {
  DesignSystemConfig,
  ConfirmationService,
  ToastService,
  Tooltip,
  CustomPreset
} from "our-library";
import "our-library/library.mjs.min.css";
import "./assets/global.scss";

createApp(App)
  .use(router)
  .use(DesignSystemConfig, {
    theme: {
      preset: CustomPreset
    }
  })
  .use(ConfirmationService)
  .use(ToastService)
  .use(createPinia())
  .directive("tooltip", Tooltip)
  .mount("#app");

```

**Reply by cristinafernandez1234** [reply]
As a workaround, we are getting the CSS variables generated in Storybook and export them and with this works, but we have to research why these CSS variables are not available in the projects.

In Storybook we have styles and variables (example from button):

<img width="302" alt="styles-storybook" src="https://github.com/user-attachments/assets/f205d080-389a-4015-bd6e-c3341e9800a8">

And in the project only styles: 

<img width="325" alt="styles" src="https://github.com/user-attachments/assets/7256585b-eff8-4b08-9b27-63c7d93ea13e">

If we export the theme and use it in the project, we have `primitive-variables` and `semantic-variables` but not the component variables:

<img width="322" alt="styles-with-theme" src="https://github.com/user-attachments/assets/14e7c051-37ed-4db9-abf0-e76f3c837c9a">

Searching about "data-primevue-style-id" we have seen another discussion that can be related: https://github.com/orgs/primefaces/discussions/2158. @vieira-brz gave different potential solutions and the one that we tried is the global style import, but we did it very manually just to test, copying each style from<style data-primevue-style-id="[COMPONENTTYPE]-variables"/> in storybook and put it in a global CSS that we export.

**Reply by Alex-Cebula** [reply]
We're having a similar issue, but oddly enough our apps that previously used our npm package with primevue v3(now updated to v4) are having no issue with the prime vue component css variables. Only our new app that is using our npm package for the first time is having an issue with undefined css variables.

**Reply by Alex-Cebula** [reply]
> We're having a similar issue, but oddly enough our apps that previously used our npm package with primevue v3(now updated to v4) are having no issue with the prime vue component css variables. Only our new app that is using our npm package for the first time is having an issue with undefined css variables.

After a few hours of digging we found that the dependencies in our package-lock.json didn't seem quite right. We deleted it and regenerated it but a minor version of Vue-tsc caused the app to fail on startup. Targeting our desired version instead of any minor version greater than it fixed this issue and now our app runs and properly populates prime vue component CSS variables even when importing a wrapped component from our npm package. Hope this helps!

**Reply by cristinafernandez1234** [reply]
Hi @Alex-Cebula, thank you for sharing your solution. I have tried also to delete the package-lock.json both in the library and in the project where the npm package is installed but I still don't have the component CSS variables, so maybe something more is missing from my side, maybe in the export script or the way the custom style is imported. Do you use style mode with custom style and you export it to use it in the project in a similar way as the following? 
```
.use(LibraryWithPrimeVue, {
  theme: {
    preset: CustomPreset
  }
})
```

**Reply by Alex-Cebula** [reply]
@cristinafernandez1234 we don't use a custom theme, we use the Lara theme and apply tweaks using design tokens and pass through options. Try taking a deeper look into your package-lock.json and make sure everything looks solid in there. We had this happen again for a new project and the culprit was a bad package-lock.

**Reply by cristinafernandez1234** [reply]
Thank you Alex, I will do so.

**Reply by johnwc** [reply]
@Alex-Cebula I do not believe a package-lock would cause any of this. A package-lock only determines what pinned version of packages are installed. You are most likely seeing it work in your main app because you are also using the components in the primary app, which causes primevue to detect the usage of them and inject the styles in the header.

If you can create a working example and publish it to a git repo that we can review it working, that would greatly help.

**Reply by Alex-Cebula** [reply]
@johnwc In our case the package-lock was causing an issue because we were using a "^" in our package.json. Which resulted in us installing a slightly newer version of prime vue than what we developed our npm package against. In some cases this caused minor inconsistencies in others it caused the absence of prime vue css variables(depends how new of a version installed). Forcing the prime vue version to match exactly between our client app and what our npm package has in its package-lock fixed this. We actually removed the "^" in both places forcing the exact version to be installed.

**Reply by johnwc** [reply]
@Alex-Cebula I think you may have found what version this bug was introduced in, is all of your packages using the latest released version of PrimeVue packages? If not, can you share what version they are pinned to, and if bumping it to the next release version after the pinned version causes it to break.

**Comment by burimdervisholli** [comment]
hi @cristinafernandez1234, any solutions yet? I kinda have the same issue looking to build a custom UI library using some of my own components and some of primevue's with a custom preset, while exporting/building the styles are missing when I try to import and use it in a test project!

**Reply by cristinafernandez1234** [reply]
No @burimdervisholli, I didn't find the solution yet. I also wrote in the discussion https://github.com/orgs/primefaces/discussions/2158 as it seems to be related. In my case the missing styles are the CSS variables for the components, so maybe there is a way to create them with a script or something (as PrimeVue does dynamically), because going component by component in Storybook and copying the variables in a CSS file is not very maintenable.

**Reply by burimdervisholli** [reply]
Thanks for the quick reply, I also do miss the css variables for the components so basically the same case. Any way lets see maybe there's a workaround for it.

**Comment by cristinafernandez1234** [comment]
Hi again! I found a workaround in this response from an issue: https://github.com/primefaces/primevue/issues/6678#issuecomment-2515992092, adding in peerDependencies: 

```
"peerDependencies": {
    "vue": "^3.2.31",
    "@primeuix/styled": "^0.0.5",
    "@primevue/themes": "^0.0.5",
    "primevue": "^4.2.2"
  }
```
